(PDB-3058) strip null bytes in terminus

wkalt opened the pull request:
Strip out null bytes in the terminus instead of in
munge-jsonb-for-storage, since null bytes will cause inserts of text to
fail in addition to jsonb.

wkalt commented:
I'm not sure this captures every representation of problematic bytes, but it at least extends our coverage of the issue to text fields. It probably negatively impacts terminus performance. We might want to have some discussion around that.

puppetlabs-jenkins commented:
Test FAILed

puppetcla commented:
CLA signed by all contributors.

ajroetker commented on puppet/lib/puppet/util/puppetdb/char_encoding.rb:
```diff
@@ -121,23 +122,26 @@ def self.coerce_to_utf8(str, error_context_str=nil)
-    if str_copy.valid_encoding?
-      str_copy.encode!("UTF-8")
+
+    if !str_copy.index("\\u0000").nil?
```
!str_copy.include?(...)

ajroetker replied:
or even
```
 str_copy.gsub!("\\u0000", DEFAULT_INVALID_CHAR) unless str_copy.include?("\\u0000")
```

ajroetker commented on puppet/lib/puppet/util/puppetdb/char_encoding.rb:
```diff
@@ -121,23 +122,26 @@ def self.coerce_to_utf8(str, error_context_str=nil)
+                                :replace => DEFAULT_INVALID_CHAR)
+
+
+    if !error_context_str.nil?
```
`unless error_context_str.nil?`

wkalt commented:
@ajroetker I fixed this up to your comments and fixed the spec test issues. I guess we're still waiting for jenkins to come back so we can run the acceptance test.

ajroetker commented:
@puppetlabs-jenkins test this please

puppetlabs-jenkins commented:
Test FAILed

ajroetker commented on acceptance/tests/inventory/basic_fact_retrieval.rb:
```diff
@@ -64,5 +64,27 @@
     end
+
+    step "create a fact with a null byte" do
+      time = Time.now.iso8691
```
You'll need to put a `require 'time'` at the top of this test

ajroetker replied:
Also @senior claims that we need 8601 here instead because ISO standard reasons

senior replied:
This function makes it so you don't have to remember the iso number https://github.com/puppetlabs/puppetdb/blob/master/puppet/lib/puppet/util/puppetdb.rb#L34

wkalt replied:
ah, yeah. 8691 is a typo. I'll fix this up.

puppetlabs-jenkins commented:
Test FAILed

wkalt commented:
@puppetlabs-jenkins retest this please

puppetlabs-jenkins commented:
Test FAILed

puppetlabs-jenkins commented:
Test FAILed

wkalt commented:
@puppetlabs-jenkins retest this please

puppetlabs-jenkins commented:
Test FAILed

puppetlabs-jenkins commented:
Test FAILed

wkalt commented:
@puppetlabs-jenkins retest this please

puppetlabs-jenkins commented:
Test FAILed

wkalt commented:
@puppetlabs-jenkins retest this please

puppetlabs-jenkins commented:
Test FAILed

puppetlabs-jenkins commented:
Test FAILed

puppetlabs-jenkins commented:
Test FAILed

mullr commented:
rebased onto 4.2.3.x, we'll see how the tests look

mullr commented:
@puppetlabs-jenkins retest this please

puppetlabs-jenkins commented:
Test FAILed

puppetcla commented:
CLA signed by all contributors.

mullr commented:
@puppetlabs-jenkins retest this please

puppetlabs-jenkins commented:
Test FAILed

mullr commented:
Closing this; we're going to try a different approach.
